src/bluealsa-dbus.c

```diff
@@ -764,6 +769,62 @@ static void bluealsa_pcm_select_codec(GDBusMethodInvocation *inv, void *userdata
 		g_variant_unref(value);
 }
 
+static void bluealsa_pcm_set_delay_adjustment(GDBusMethodInvocation *inv, void *userdata) {
+
+	GVariant *params = g_dbus_method_invocation_get_parameters(inv);
+
+	struct ba_transport_pcm *pcm = (struct ba_transport_pcm *)userdata;
+
+	const char *codec_name;
+	int16_t adjustment;
+
+	g_variant_get(params, "(&sn)", &codec_name, &adjustment);
+	uint16_t codec_id = 0xFFFF;
+	bool codec_name_valid = false;
+	if (pcm->t->profile & BA_TRANSPORT_PROFILE_MASK_A2DP) {
+		codec_id = a2dp_codecs_codec_id_from_string(codec_name);
+		codec_name_valid = codec_id != 0xFFFF;
+	}
+	if (pcm->t->profile & BA_TRANSPORT_PROFILE_MASK_SCO) {
+		codec_id = hfp_codec_id_from_string(codec_name);
+		codec_name_valid = codec_id != HFP_CODEC_UNDEFINED;
+	}
+	if (codec_name_valid) {
+		ba_transport_pcm_delay_adjustment_set(pcm, codec_id, adjustment);
+		bluealsa_dbus_pcm_update(pcm, BA_DBUS_PCM_UPDATE_DELAY_ADJUSTMENT);
+	}
+	else {
+		error("Invalid codec name (%s)", codec_name);
+	}
+	g_dbus_method_invocation_return_value(inv, NULL);
+}
+
+static void bluealsa_pcm_get_delay_adjustments(GDBusMethodInvocation *inv, void *userdata) {
+
+	struct ba_transport_pcm *pcm = (struct ba_transport_pcm *)userdata;
+	const struct ba_transport *t = pcm->t;
+
+	GVariantBuilder adjustments;
+	g_variant_builder_init(&adjustments, G_VARIANT_TYPE("a{sn}"));
+
+	GHashTableIter iter;
+	g_hash_table_iter_init(&iter, pcm->delay_adjustments);
+
+	gpointer key, value;
+	while (g_hash_table_iter_next(&iter, &key, &value)) {
```

Comment:
Iteration here needs a mutex to protect against race with `bluealsa_pcm_set_delay_adjustment()`. All BlueALSA D-Bus API calls are handled in a separated thread, so they can run in parallel.